pkg/reconciler/app/reconciler.go

```diff
@@ -328,49 +329,52 @@ func (r *Reconciler) ApplyChanges(ctx context.Context, app *v1alpha1.App) error
 	{
 		logger.Debug("reconciling Service Account")
 		condition := app.Status.ServiceAccountCondition()
-
-		// Fetch the secrets referenced in the build service account, and filter out the token secret auto-generated by K8s.
-		// If Workload Identity is enabled, there should be no secrets after filtering.
-		// If WI is not enabled, then the secrets should have the same values as those from build.ImagePushSecrets in config-secrets.
-		// These secrets will be passed into ImagePullSecrets on the App Service Account.
-		// ImagePullSecrets are only accessed by the kubelet and are not mounted in the Pod.
-		buildSA, err := r.serviceAccountLister.ServiceAccounts(app.Namespace).Get(spaces.BuildServiceAccountName(space))
+		cfg, err := kfconfig.FromContext(ctx).Defaults()
 		if err != nil {
-			return condition.MarkReconciliationError("getting build SA", err)
+			return fmt.Errorf("couldn't read config-defaults from context: %w", err)
 		}
 
-		buildSecrets := []*v1.Secret{}
-		for _, s := range buildSA.Secrets {
-			secret, err := r.SecretLister.Secrets(app.Namespace).Get(s.Name)
-			if err != nil {
-				return condition.MarkReconciliationError("getting secret", err)
-			}
-			buildSecrets = append(buildSecrets, secret)
+		overrideKsaName := ""
+		if annotations := app.ObjectMeta.Annotations; len(annotations) > 0 {
+			overrideKsaName = annotations[serviceAccountNameOverrideAnnotation]
 		}
 
-		filteredSecretNames := []v1.LocalObjectReference{}
-		for _, s := range resources.FilterAndSortKfSecrets(buildSecrets) {
-			filteredSecretNames = append(filteredSecretNames, v1.LocalObjectReference{Name: s.Name})
-		}
+		switch {
```

Comment:
nit: why a switch here vs if / else?

Reply:
I think that's valid although I prefer using switch whenever there isn't a clear boolean test. e.g. in this case, I think it would be perfectly valid to throw in another case where a user tries to override a KSA but it isn't enabled that logs an error and falls through or where an operator could override this for the whole space. In my mind it's similar to using enums rather than booleans if there's the slightest chance there might be more.

The block farther down has always bugged me because go doesn't allow assignment in case statements so we have that ugly if/else if/else chain.